kwok/cloudprovider/cloudprovider.go

```diff
@@ -0,0 +1,271 @@
+	//nolint
+	newName = fmt.Sprintf("%s-%d", newName, rand.Uint32())
+
+	capacityType := v1beta1.CapacityTypeOnDemand
+	requirements := scheduling.NewNodeSelectorRequirements(nodeClaim.
+		Spec.Requirements...)
+	if requirements.Get(v1beta1.CapacityTypeLabelKey).Has(v1beta1.CapacityTypeSpot) {
+		capacityType = v1beta1.CapacityTypeSpot
+	}
+	req, found := lo.Find(nodeClaim.Spec.Requirements, func(req v1.NodeSelectorRequirement) bool {
+		return req.Key == v1.LabelInstanceTypeStable
+	})
+	if !found {
+		return nil, fmt.Errorf("instance type requirement not found")
+	}
+
+	minInstanceTypePrice := math.MaxFloat64
+	var instanceType *cloudprovider.InstanceType
+	// Loop through instance type values, as the node claim will only have the In operator.
+	for _, val := range req.Values {
+		var price float64
+		var ok bool
+		it, err := c.getInstanceType(val)
+		if err != nil {
+			return nil, fmt.Errorf("instance type %s not found", val)
+		}
+
+		// Since we're constructing the instance types we know that each instance type with OD offerings will have spot
+		// offerings, where spot will always be cheapest.
+		zone := randomChoice(KwokZones)
+		offering, ok := it.Offerings.Get(capacityType, zone)
+		if !ok {
+			return nil, fmt.Errorf("failed to find offering %s/%s/%s", capacityType, zone, val)
+		}
+		// All offerings of the same capacity type are the same price.
+		price = offering.Price
+		if price < minInstanceTypePrice {
+			minInstanceTypePrice = price
+			instanceType = it
+		}
+	}
+
+	return &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        newName,
+			Labels:      addInstanceLabels(nodeClaim.Labels, instanceType, nodeClaim, capacityType, fmt.Sprintf("%f", minInstanceTypePrice)),
+			Annotations: addKwokAnnotation(nodeClaim.Annotations),
+		},
+		Spec: v1.NodeSpec{
+			ProviderID: kwokProviderPrefix + newName,
+		},
+		Status: v1.NodeStatus{
+			Capacity:    instanceType.Capacity,
+			Allocatable: instanceType.Allocatable(),
+			Phase:       v1.NodePending,
+		},
+	}, nil
+}
+
+func addInstanceLabels(labels map[string]string, instanceType *cloudprovider.InstanceType, nodeClaim *v1beta1.NodeClaim, capacityType, price string) map[string]string {
+	ret := make(map[string]string, len(labels))
+	// start with labels on the machine
+	for k, v := range labels {
+		ret[k] = v
+	}
+
+	// add the derived machine requirement labels
+	for _, r := range nodeClaim.Spec.Requirements {
+		if len(r.Values) == 1 && r.Operator == v1.NodeSelectorOpIn {
+			ret[r.Key] = r.Values[0]
+		}
+	}
+
+	// ensure we have an instance type and then any instance type requirements
+	ret[v1.LabelInstanceTypeStable] = instanceType.Name
+	for _, r := range instanceType.Requirements {
+		if r.Len() == 1 && r.Operator() == v1.NodeSelectorOpIn {
+			ret[r.Key] = r.Values()[0]
+		}
+	}
+	// add in github.com/awslabs/eks-node-viewer label so that it shows up.
+	ret["eks-node-viewer/instance-price"] = price
```

Comment:
nit: reference a constant value? Either in the package itself or in our own constants file?